File: 28. Find the Index of the First Occurrence in a String/find-the-index-of-the-first-occurrence-in-a-string.py
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        def build_lps(s):
            lps = [0] * len(s)
            i = 1
            l = 0
            while i < len(s):
                # I already know the current substring ends with a prefix of length l. 
                # Can I make that prefix one character longer?
                if s[i] == s[l]:
                    l += 1
                    lps[i] = l
                    i += 1
                elif l > 0:
                    l = lps[l - 1]
                else:
                    lps[i] = 0
                    i += 1
            return lps
        # Knuth–Morris–Pratt (KMP)
        # Time: O(n + m)
        # After a mismatch, move j to lps[j - 1]
        # ABABA|C|   ->    ABA|B|AC
        # ABABA|B|C  ->  ABABA|B|C
        lps = build_lps(needle)
        i, j = 0, 0
        while i < len(haystack):
            if haystack[i] == needle[j]:
                i += 1
                j += 1
                if j == len(needle): return i - j
            elif j != 0:
                j = lps[j - 1]
            else:
                i += 1
        return -1
            

        
    

File: 28. Find the Index of the First Occurrence in a String/find-the-index-of-the-first-occurrence-in-a-string.java
class Solution {
    public static int[] buildLPS(String s) {
        int[] lps = new int[s.length()];
        int i = 1;
        int len = 0;
        while (i < s.length()) {
            if (s.charAt(i) == s.charAt(len)) {
                len++;
                lps[i] = len;
                i++;
            } else if (len > 0) {
                len = lps[len - 1];
            } else {
                lps[i] = 0;
                i++;
            }
        }
        return lps;
    } 
    public int strStr(String haystack, String needle) {
        int[] lps = buildLPS(needle);
        int j = 0;
        int i = 0;
        while (i < haystack.length()) {
            if (haystack.charAt(i) == needle.charAt(j)) {
                i++;
                j++;
                if (j == needle.length()) return i - j;
            } else if (j > 0) {
                j = lps[j - 1];
            } else {
                i++;
            }
        } 
        return -1;
    }
}
